engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java

```diff
@@ -93,188 +71,126 @@ public TFloatList getVertices() {
-            parts.add(newData.getTexCoord1().iterator());
-            partSizes.add(TEX_COORD_1_SIZE);
-            texCoord1Offset = vertexSize * FLOAT_SIZE;
-            vertexSize += TEX_COORD_1_SIZE;
-            hasTexCoord1 = true;
-        }
-        if (newData.getNormals() != null && newData.getNormals().size() / NORMAL_SIZE == vertexCount) {
-            parts.add(newData.getNormals().iterator());
-            partSizes.add(NORMAL_SIZE);
-            normalOffset = vertexSize * FLOAT_SIZE;
-            vertexSize += NORMAL_SIZE;
-            hasNormal = true;
-        }
-        if (newData.getColors() != null && newData.getColors().size() / COLOR_SIZE == vertexCount) {
-            parts.add(newData.getColors().iterator());
-            partSizes.add(COLOR_SIZE);
-            colorOffset = vertexSize * FLOAT_SIZE;
-            vertexSize += COLOR_SIZE;
-            hasColor = true;
-        }
-        stride = vertexSize * FLOAT_SIZE;
-        indexCount = newData.getIndices().size();
+        List<Layout> layouts = newData.getLayouts();
+        List<Layout> targets = new ArrayList<>();
 
-        createVertexBuffer(parts, partSizes, vertexCount, vertexSize);
-        createIndexBuffer(newData.getIndices());
-        getBound(newData, aabb);
-    }
+        int stride = 0;
+        for (Layout layout : layouts) {
+            if (layout.hasContent()) {
+                stride += layout.bytes();
+                targets.add(layout);
+            }
+        }
 
-    private void createVertexBuffer(List<TFloatIterator> parts, TIntList partSizes, int vertexCount, int vertexSize) {
-        FloatBuffer vertexBuffer = BufferUtils.createFloatBuffer(vertexSize * vertexCount);
-        for (int v = 0; v < vertexCount; ++v) {
-            for (int partIndex = 0; partIndex < parts.size(); ++partIndex) {
-                TFloatIterator part = parts.get(partIndex);
-                for (int i = 0; i < partSizes.get(partIndex); ++i) {
-                    vertexBuffer.put(part.next());
-                }
+        ByteBuffer buffer = BufferUtils.createByteBuffer(stride * newData.getSize());
+        for (int x = 0; x < newData.getSize(); x++) {
+            for (Layout layout : targets) {
+                layout.write(x, buffer);
             }
         }
-        vertexBuffer.flip();
-        if (disposalAction.vboVertexBuffer == 0) {
-            disposalAction.vboVertexBuffer = disposalAction.bufferPool.get(getUrn().toString());
+        buffer.flip();
+
+        // bind vertex array and buffer
+        GL30.glBindVertexArray(this.disposalAction.vao);
+        GL30.glBindBuffer(GL30.GL_ARRAY_BUFFER, this.disposalAction.vbo);
+        GL30.glBufferData(GL30.GL_ARRAY_BUFFER, buffer, GL30.GL_STATIC_DRAW);
+
+        int offset = 0;
+        for (Layout layout : targets) {
+            GL30.glEnableVertexAttribArray(layout.location);
+            if ((layout.flag & Layout.FLOATING_POINT) > 0) {
```

Comment:
I have some plans to do a second pass on this. I want to split the layout into a nested class within a definition. this is actually better for memory performance: https://anteru.net/blog/2016/storing-vertex-data-to-interleave-or-not-to-interleave/ I was also looking a bit UnrealEngine to see how it manages buffers and they keep things non interleaved. 

Vertex definition would contain the memory and the attribute will define how the data is mapped into a buffer. we can have version flag so only the vertex data that is changed will be mapped into the VBO. I want to rework this a again with some research I've been doing on the side as I was setting this up. we have to handle the case where the size of the definition changes so all the buffered data will have to be re-written. still debating if it would be better to have one VBO or multiple VBO's. GL30 has a subdata function where you can save into part of the buffer and not attempt to overwrite the whole thing.  One thing I want to do in particular is have the data written straight to the buffer and I think we could have a flag that would let us save some of the data to some kind of backing array if that is important. 

- VertexDefintion (Position Color)
  - VertexAttribute (Position)
  - VertexAttribute (Color)
- VertexDefintion (UV0) 
  - VertexAttribute () 
- VertexDefintion (UV1)
  - VertexAttribute

Reply:
I tried to test this but it turned out to be a huge pain to set up new shaders, a new material, a new sort of mesh and a prefab to use that mesh, just to get one integer attribute, so I gave up. The code looks correct, so I'll just accept it without testing this bit.